### ISSUE TITLE
Lightmapping, is it time to consider this option?

### ISSUE THREAD

**Issue by davetgc** [TheGameCreators/GameGuruRepo#4074]
I've seen several uses mention the limitation of dynamic light mapping across the board, and as we know,it can have a performance hit. 

Should we be considering a static light system as well as a dynamic light, where shadows are pre-rendered to a lightmap? There dynamic lights could be used more selectively and to greater effect.

Thoughts welcome?

**Comment by MonkeyFrogStudio** [comment]
Lightmaps should never auto-generate (in my opinion). Even Unreal lets you know that the lighting has changed and your lightmap needs to be updated instead of just auto-generating a new one. So, yes, lightmaps would be great for many situations, and no, they should not just autogenerate every time someone starts a test game.

Secondly, since the viewport shows us what we see in game graphically (i.e. lights cast shadows in the viewport, etc.), then we should be able to generate and update the lightmap and see it in the viewport, too. Again, engines like Unreal have this feature as standard (that you generate the lightmap and can see the end result in the editor).

**Comment by ShadowMasterOfGames** [comment]
Even wicked engine has that feature the @MonkeyFrogStudio mention above

**Comment by synchromesh62** [comment]
Ironically i just did a vid about all of this ..
Max allows 16 lights to cast shadows in one area ( distance detected ) so you can mix and match lights with enabling **" Do not cast shadow "** on many you really do not need to use with shadows.
As soon as you move far enough away from the area the shadows are turned off and enabled in the new area so no matter how big your map is the most your using is 16 lights with 16 shadows as this vid shows ... The first section has 18 lights and shows the 2 extra being enabled as i get close ... As you leave the area and enter another you have a new set of 16 lights casting shadows .. 
https://youtu.be/Npx9JM_AIc0

**Comment by Kravenwolf** [comment]
I imagine static lights would offer one hell of an optimization for VR games. Pre-calculating static lightmaps instead of doing it twice every frame in real-time for dynamic ones (that is how VR works, yes?).

In any case, I'm a strong believer that ANY feature that is a part of Classic should be carried over into MAX, as was the case with SSAO. There shouldn't be a single case where someone using MAX is looking for a feature from it's 2015 predecessor and realizes it's not an option for them. Nothing "MAX" about that at all.

**Comment by davetgc** [comment]
I'm putting this out toe @LeeBamberTGC  and @ZakJudges . We already have a similar system in Classic, how portable is that?

**Comment by MonkeyFrogStudio** [comment]
@davetgc Wicked has a lightmapper. At least it did. Even a voxel lightmapper. It seems to me that it would make sense to wait until the DX12/Wicked integration is finished, then we'd have access to Wicked's lightmapper in MAX.

**Comment by synchromesh62** [comment]
Have not really ever used the lightmapper in Classic.
But from what i remember shadows are static which was fine foe Classic as it did not have Dynamic shadows at all.
But wont that just mess up the Realtime shadows? Picking up a box, using a flashlight ?
If not and the shadows are still active, what would be the point ?

**Comment by davetgc** [comment]
No, as all non static objects, will use the realtime lighting It would only be static objects that would use the lightmapper. It's possible this could mess with a day and night system, but IMO the benefits outweigh the disadvantages.

**Comment by Granada1** [comment]
Great for cutting down on real-time lights in a level

**Comment by synchromesh62** [comment]
> No, as all non static objects, will use the realtime lighting It would only be static objects that would use the lightmapper. It's possible this could mess with a day and night system, but IMO the benefits outweigh the disadvantages.

Ahh ok i just had visions of the flashlight casting other shadows but still seeing the lightmapped ones etc which to me seemed would look a bit odd plus to cast realtime shadows indoors for movable objects you need a dynamic light which again could look off. Im probably way off but just seemed logical to me that would be the effect when mixing.

**Comment by davetgc** [comment]
My thinking would be that maybe realtime shadows are used close up, for things like flashlight, day/night, etc, with the baked map used further out. I'm only really putting out ideas atm.

**Comment by OldFlak** [comment]
Please don't use GGc's light mapper - it never worked properly, not to mention the bloated fpm as a result, and in most cases there would be somewhere on the map that the LM destroyed.

If wicked has one, why not use that?
